zerver/lib/actions.py

```diff
@@ -2776,6 +2778,24 @@ def user_info(um):
     send_event(event, list(map(user_info, ums)))
     return len(changed_messages)
 
+# Calculate the average weekly traffic of a stream
+def get_stream_message_count(stream_id, date_created):
+    # type: (int, datetime.datetime) -> int
+    count = 0
+    stat = COUNT_STATS['messages_in_stream:is_bot:day']
+    d = timezone.now() - datetime.timedelta(days=28)
+    queryset = StreamCount.objects.filter(stream_id=stream_id, property=stat.property,
+                                          end_time__gt=d).values_list('value')
+    for val in queryset:
+        count += val[0]
+    num_days = (timezone.now().date() - date_created.date()).days
+    # Calculates the average over 4 weeks
+    # If duration less than 4 weeks, adjust accordingly
+    if num_days > 6:
+        count = int(count // 4) if num_days >= 28 else int(count // (num_days // 7))
+    # count is rounded off to 2 significant figures
+    return int(round(count, 2-len(str(count))))
+
```

Comment:
So, the reason the tests are failing is because you're doing database queries in a loop here inside `gather_subscriptions_helpers`, which would cause a big performance problem.

What we want to do instead in a single bulk query on the `StreamCount` table, and splice the data in there.